### ISSUE TITLE
Add "effective fee rate" in rpc's mempool entries

### ISSUE THREAD

**Issue by RHavar** [bitcoin/bitcoin#10389]
For both `getrawmempool true` and `getmempoolentry` it should probably have a field `effectivefeerate`.

Where as "effective fee rate" is defined as what miners sort transactions by (when taking into consideration the whole CPFP stuff).

I need this field for when trying to create my own CPFP transactions, but it's quite difficult to calculate myself (especially as the mempool entries don't have a "dependants" field, so you need to dump the whole mempool.

**Comment by sipa** [comment]
You can divide `ancestorfees` divided by `ancestorsize` to get the feerate in sat/byte for a transaction including all its dependencies.

**Comment by RHavar** [comment]
@sipa  That doesn't seem correct


Assume all transactions are the same size, and we have 4 unconf transactions (A, B, C and D):

A  = 10 satoshis of fees
B  = 100 satoshis of fees  (and depends on A)
C  = 1  satoshi of fees (and depends on B)
D = 1000 satoshis of fees (and depends on B)

---

Now, if we're trying to find the effective fee rate of transaction B, we need the fees/size of (A, B, D). A we care about as it's a dependency.  We don't care about C because it's not relevant. And we care about D because it boosts the effective fee rate of the transaction "cluster".

**Comment by sdaftuar** [comment]
@rhavar I think you may be interested in the `getmempoolancestors` and `getmempooldescendants` rpc calls (introduced in https://github.com/bitcoin/bitcoin/pull/7292).

>Assume all transactions are very small, and we have 4 unconf transactions (A, B, C and D):
>
>A = 10 satoshis of fees
>B = 100 satoshis of fees (and depends on A)
>C = 1 satoshi of fees (and depends on B)
>D = 1000 satoshis of fees (and depends on B)

Keep in mind that it's a bit tricky to exactly calculate the "effective" feerate of transaction B.  Imagine if A had some other descendant, E, which has a much higher feerate than D.  Then A's effective feerate is actually determined by E, and B's effective feerate is determined only by B and D (since B no longer has to pay for A).

If you're interested in transaction B and willing to ignore the effect of descendants of B's ancestors on the effective feerate of B, then I think you could just look at all the descendants of B, and look at the max ancestor feerate of B and each of B's descendants.  Does that seem like it'd be good enough for your use cases?

**Comment by RHavar** [comment]
> @RHavar I think you may be interested in the getmempoolancestors and getmempooldescendants rpc calls (introduced in #7292).

Yeah, that's neat. It'll make my code a little simpler, I won't need to dump the whole mempool. It's not ideal though, because I still need to:

getmempoolancestors on the transaction, and then loop over all the results with getmempooldescendants  to compile enough information.

> Keep in mind that it's a bit tricky to exactly calculate the "effective" feerate of transaction B. 

Yeah, it's annoying. That's why it should be provided by bitcoin :P

> Does that seem like it'd be good enough for your use cases?

Nope :(  I really need to know exactly in order to do CPFP myself properly.

**Comment by glozow** [comment]
Seems related to #15553 and #9645.
Currently working on this.

**Comment by willcl-ark** [comment]
Ping @murchandamus, does this work with MiniMiner now?

**Comment by pinheadmz** [comment]
Is this closed by cluster mempool overhaul?

```
 bitcoin-cli getmempoolentry 96a6ac96cdb7d03fa45dc56cfca9ca05887000575e4ba8cc5f43bd50ef3b3e38
{
  "vsize": 95476,
  "weight": 381902,
  "time": 1777859488,
  "height": 947805,
  "descendantcount": 1,
  "descendantsize": 95476,
  "ancestorcount": 1,
  "ancestorsize": 95476,
  "wtxid": "3f558448fe30ed5744f86ce403a99a138d4e5b88cf0b21fba55cee683d37b424",
  "chunkweight": 381902,
  "fees": {
    "base": 0.00009548,
    "modified": 0.00009548,
    "ancestor": 0.00009548,
    "descendant": 0.00009548,
    "chunk": 0.00009548
  },
  "depends": [
  ],
  "spentby": [
  ],
  "bip125-replaceable": true,
  "unbroadcast": false
}
```

**Comment by glozow** [comment]
Yes, can be closed as completed

**Comment by pinheadmz** [comment]
❤️ ❤️ ❤️
